src/services/redisSessionStorage.ts

```diff
@@ -31,35 +42,89 @@ export default class RedisSessionStorage implements ISessionStorage {
+        token,
+        REDIS_DATA_FIELD,
+        REDIS_TIMESTAMP_FIELD,
+        (err, value) => {
+          if (err) {
+            resolve(left<Error, User>(new Error(err.message)));
           } else {
-            const errorOrUser = extractUserFromJson(value);
+            if (value === null || value === undefined) {
+              resolve(left<Error, User>(new Error(SESSION_NOT_FOUND_MESSAGE)));
+            } else {
+              // check if the token has expired. We don't remove expired token
+              // because client can later refresh the session.
+              if (+value[1] + this.tokenDuration * 1000 < Date.now()) {
+                resolve(left<Error, User>(new Error("Token has expired")));
+              }
+
+              const errorOrUser = extractUserFromJson(value[0]);
+
+              // TODO: better error message.
+              errorOrUser.mapLeft(() => {
+                return "Errors in validating the user profile";
+              });
+
+              resolve(errorOrUser);
+            }
+          }
+        }
+      );
+    });
+  }
 
-            // TODO: better error message.
-            errorOrUser.mapLeft(() => {
-              return "Errors in validating the user profile";
-            });
+  /**
+   * {@inheritDoc}
+   */
+  public refresh(token: string): Promise<Either<Error, string>> {
+    const client = this.redisClient;
 
-            resolve(errorOrUser);
+    return new Promise(resolve => {
+      // Get the fields for this key.
+      // @see https://redis.io/commands/hmget
+      client.hmget(
+        token,
+        REDIS_DATA_FIELD,
+        REDIS_TIMESTAMP_FIELD,
+        (err, value) => {
+          if (err) {
+            resolve(left<Error, string>(new Error(err.message)));
+          } else {
+            if (value === null || value === undefined) {
+              resolve(
+                left<Error, string>(new Error(SESSION_NOT_FOUND_MESSAGE))
+              );
+            } else {
+              const errorOrUser = extractUserFromJson(value[0]);
+              errorOrUser.fold(
+                () => {
+                  resolve(
+                    left<Error, string>(new Error(SESSION_NOT_FOUND_MESSAGE))
+                  );
+                },
+                (user: User) => {
+                  const newToken = this.tokenService.getNewToken();
+                  this.set(newToken, user);
+                  this.del(token);
+                  resolve(right<Error, string>(newToken));
+                }
+              );
+            }
           }
         }
-      });
+      );
     });
   }
 
   /**
    * {@inheritDoc}
    */
   public del(token: string): void {
-    // Removes the specified keys. A key is ignored if it does not exist.
+    // Remove the specified key. A key is ignored if it does not exist.
     // @see https://redis.io/commands/hdel
-    this.redisClient.hdel("sessions", token);
+    this.redisClient.hdel(token, REDIS_DATA_FIELD, REDIS_TIMESTAMP_FIELD);
```

Comment:
is this function synchronous? if not (and I hope so), if should have a callback...